src/boot.c

```diff
@@ -451,9 +452,10 @@ void read_boot(DOS_FS * fs)
     struct boot_sector b;
     unsigned total_sectors;
     unsigned int logical_sector_size, sectors;
-    off_t fat_length;
+    long long fat_length;
     unsigned total_fat_entries;
     off_t data_size;
+    long long position;
```

Comment:
It should be unsigned as you are doing assignment from unsigned values.

Reply:
In fact, the assignments are already casted to long long explicitly or implicitly. I see no concern here because the data type is large enough to hold all possible values.

Reply:
You are doing assignment from `((total_sectors & ~1) - 1) * logical_sector_size` and this is always unsigned. So I do not think there is need to store non-negative value into signed type if there is no need to represent negative values.

Reply:
So you prefer to have an unsigned position which is used to assign off_t's later on? I don't get the point here.

Reply:
The result of `((total_sectors & ~1) - 1)` is explicitly casted to `(long long)` which is save as long as unsigned is smaller than long long in terms of available bits. The multiplication with logical_sector_size is therefore signed and long long as well.

Reply:
Hm... I'm thinking about it... `total_sectors` is unsigned and maximally 32bit. `logical_sector_size` is unsigned and maximally 16bit. Result is unsigned, so should be stored in `unsigned long long`. But also it always fit into 24bits, therefore also `signed long long` type is suitable (after casting). But you are right that `off_t` is signed and in most cases up to the `2^63-1` or `2^31-1`.

Personally I prefer to not use "less" ambiguous types (like `off_t` or `time_t`) and rather `unsigned long long` which is guaranteed to be at least 64 bit long.

And in case that value must not be negative (like number of sectors, sector size or offset from beginning of file) then I rather prefer to use unsigned type to enforce this constraint. Usage of signed type tells reader "this is signed type, it can be negative, so be careful".

Reply:
I agree here and I would have used uint64_t for that (and later on switched to unsigned long long) because negative values wouldn't make any sense for these values, after all they are unsigned in FAT specification as well.

But due to usage of off_t throughout the code and especially for storing these values in structs I've kept using a signed type. In fact, on linux/amd64 these changes are basically no-ops because off_t == long long.

This keeps my changes as small as possible in terms of "change of behaviour" on amd64 and increases the width of data types far enough on these few systems with off_t being 32 bit, i.e. linux/i686 and others with --disable-largefile.

Reply:
Probably the best fix would be to use unsigned values in structs and convert them to `off_t` just before passing them into `lseek` function.

So we can let decision to @andreasbombe how he wants to have this problem handled/fixed.

Current usage of types is really mess and I understand that you want ot keep changes as small as possible.

Reply:
Yes, that sounds great. In that case read_boot can fully validate that the boot sector is correct even on systems that couldn't handle the file system later on due to size limitations.

And such a change should definitely go into its own merge request...